src/cc65/codegen.c

```diff
@@ -2590,13 +2550,26 @@ void g_mul (unsigned flags, unsigned long val)
         "tosmulax", "tosumulax", "tosmuleax", "tosumuleax"
     };
 
-    int p2;
-
     /* Do strength reduction if the value is constant and a power of two */
-    if (flags & CF_CONST && (p2 = PowerOf2 (val)) >= 0) {
-        /* Generate a shift instead */
-        g_asl (flags, p2);
-        return;
+    if (flags & CF_CONST) {
+
+        /* Deal with negative values if it's signed multiplication */
+        int Negation = (flags & CF_UNSIGNED) == 0 && (signed long)val < 0;
+        int p2 = PowerOf2 (Negation ? (unsigned long)-(signed long)val : val);
```

Comment:
Style nit: to me, `(0 - val)` is clearer and requires less thinking than `(unsigned long)-(signed long)val`.